File: py-algorithms/hckrrnk/quick_sort.py


def quick_sort(n, arr):
    pivot = arr[0]
    e = []
    l = []
    r = []

    for i in range(n):
        if arr[i] < pivot:
            l.append(arr[i])
            print(l)
        if arr[i] > pivot:
            r.append(arr[i])
            print(r)
        if arr[i] == pivot:
            e.append(arr[i])
   
    res = l + e + r
    
    return res

if __name__ == '__main__':
    print(quick_sort(5, [4,5,3,7,2]))

File: go-tryouts/hckrnk/quick_sort.go
package hckrnk

import "log"

func QuickSort(n int, arr []int) []int {
	pivot := arr[0]
	e := make([]int, 0)
	l := make([]int, 0)
	r := make([]int, 0)
	helpers := make([][]int, 0)
	res := make([]int, 0)

	for i, _ := range arr {
		if arr[i] < pivot {
			l = append(l, arr[i])
		}
		if arr[i] > pivot {
			r = append(r, arr[i])
		}
		if arr[i] == pivot {
			e = append(e, arr[i])
		}
	}
	// concatenate slices
	helpers = append(helpers, l, e, r)
	log.Println(helpers)
	for _, v := range helpers {
		// fmt.Println(v)
		res = append(res, v...)

	}
	// fmt.Println(res)

	return res
}
